userland/aero_ipc/src/lib.rs

```diff
@@ -0,0 +1,223 @@
+/*
+ * Copyright (C) 2021-2022 The Aero Project Developers.
+ *
+ * This file is part of The Aero Project.
+ *
+ * Aero is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Aero is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Aero. If not, see <https://www.gnu.org/licenses/>.
+ */
+#![feature(decl_macro)]
+
+use aero_syscall::{sys_ipc_recv, sys_ipc_send};
+use core::ops::DerefMut;
+use core::sync::atomic::{AtomicUsize, Ordering};
+use lazy_static::lazy_static;
+use serde::Deserialize;
+
+/// This is an ugly hack that exposes postcard and serde for the macro to use
+pub mod ipcmodules {
+    pub use postcard;
+    pub use serde;
+}
+
+/// A MessageHandler is a trait describing an IPC client
+pub trait MessageHandler: Send + Sync {
+    fn handle(&mut self, src: usize, msg: &[u8]) -> Option<Vec<u8>>;
+}
+
+/// A MessageTransport allows for high-level IPC exchanges over the IPC interfce.
+/// It also handles the allocation of request identifiers.
+pub trait MessageTransport {
+    fn alloc_id() -> usize;
+    fn free_id(id: usize);
+    fn exchange(meta: usize, mid: usize, data: &[u8]) -> Vec<u8>;
+}
+
+/// A SendRecieveTransport transfers messages by using the IPC system calls.
+pub struct SendRecieveTransport;
+static IDALLOC: AtomicUsize = AtomicUsize::new(0);
```

Comment:
Personally I think it would be better for the kernel to take care of allocating message ID's for the messages in some way or another, just to rule out possible malicious use cases where you forge messages with given message ID but I'm unsure how all of this interacts with each other so take it with a grain of salt

Reply:
to prevent response forging, i think i will just check that the request comes from the correct process, and possibly randomize the IDs too (just as an extra mitigation).

Reply:
Sure, I think we can roll with that for now thought as it's good enough :+1:

Reply:
fixed it anyway, just for good measure.